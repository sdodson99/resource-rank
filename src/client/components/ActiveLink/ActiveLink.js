import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import { useRouter } from 'next/router';

const ActiveLink = ({ href, className, activeClassName, children }) => {
  const router = useRouter();

  if (href === router?.asPath) {
    return (
      <div
        className={`${className} ${activeClassName}`}
        data-testid="ActiveLink"
      >
        {children}
      </div>
    );
  }

  return (
    <div className={className} data-testid="ActiveLink">
      <Link href={href}>
        <a>{children}</a>
      </Link>
    </div>
  );
};

ActiveLink.propTypes = {
  href: PropTypes.string.isRequired,
  className: PropTypes.string,
  activeClassName: PropTypes.string,
  children: PropTypes.node,
};

ActiveLink.defaultProps = {
  href: '#',
  className: '',
  activeClassName: 'active',
};

export default ActiveLink;
