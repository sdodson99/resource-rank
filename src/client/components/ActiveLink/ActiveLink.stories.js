import React from 'react';
import ActiveLink from './ActiveLink';

export default {
  component: ActiveLink,
  title: 'Components/ActiveLink',
  decorators: [
    (Story) => {
      return <Story />;
    },
  ],
};

const Template = (args) => <ActiveLink {...args} />;

export const Default = Template.bind({});
Default.args = {
  href: '/page',
  children: 'Click me',
};

export const Active = Template.bind({});
Active.args = {
  href: '',
  children: 'Already active',
  activeClassName: 'text-red-800',
};
