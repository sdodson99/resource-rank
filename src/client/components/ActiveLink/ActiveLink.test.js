import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ActiveLink from './ActiveLink';
import { createRenderer } from 'react-test-renderer/shallow';
import { useRouter } from 'next/router';

jest.mock('next/router');

describe('<ActiveLink />', () => {
  let props;

  beforeEach(() => {
    props = {
      url: {
        pathname: '/',
      },
      className: 'className',
      linkClassName: 'linkClassName',
    };
  });

  afterEach(() => {
    useRouter.mockReset();
  });

  it('should mount', () => {
    render(<ActiveLink {...props}>Link</ActiveLink>);

    const activeLink = screen.getByTestId('ActiveLink');

    expect(activeLink).toBeInTheDocument();
  });

  it('should render correctly', () => {
    const page = createRenderer().render(
      <ActiveLink {...props}>Link</ActiveLink>
    );

    expect(page).toMatchSnapshot();
  });

  describe('with active', () => {
    let path;

    beforeEach(() => {
      path = '/home';

      props.url.pathname = path;
      useRouter.mockReturnValue({ asPath: path });
    });

    it('should render correctly', () => {
      const page = createRenderer().render(
        <ActiveLink {...props}>Link</ActiveLink>
      );

      expect(page).toMatchSnapshot();
    });

    it('should have active class name', () => {
      render(<ActiveLink {...props}>Link</ActiveLink>);

      const activeLink = screen.getByTestId('ActiveLink');

      expect(activeLink.classList).toContain('active');
    });
  });

  describe('with active and query parameters', () => {
    let path;

    beforeEach(() => {
      path = '/home';
      props.url.pathname = path;

      useRouter.mockReturnValue({ asPath: '/home?test=123' });
    });

    it('should render correctly', () => {
      const page = createRenderer().render(
        <ActiveLink {...props}>Link</ActiveLink>
      );

      expect(page).toMatchSnapshot();
    });

    it('should have active class name', () => {
      render(<ActiveLink {...props}>Link</ActiveLink>);

      const activeLink = screen.getByTestId('ActiveLink');

      expect(activeLink.classList).toContain('active');
    });
  });

  describe('with not active', () => {
    beforeEach(() => {
      useRouter.mockReturnValue({ asPath: '/home' });
    });

    it('should render correctly', () => {
      const page = createRenderer().render(
        <ActiveLink {...props}>Link</ActiveLink>
      );

      expect(page).toMatchSnapshot();
    });

    it('should have active class name', () => {
      render(<ActiveLink {...props}>Link</ActiveLink>);

      const activeLink = screen.getByTestId('ActiveLink');

      expect(activeLink.classList).not.toContain('active');
    });
  });
});
