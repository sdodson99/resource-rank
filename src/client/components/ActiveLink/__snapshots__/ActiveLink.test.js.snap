// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActiveLink /> with active should render correctly 1`] = `
<div
  className=" active"
  data-testid="ActiveLink"
>
  Link
</div>
`;

exports[`<ActiveLink /> with not active should render correctly 1`] = `
<div
  className=""
  data-testid="ActiveLink"
>
  <Link
    href="/"
  >
    <a>
      Link
    </a>
  </Link>
</div>
`;
